src/lib/core/style/add-focus-classes.ts

```diff
@@ -0,0 +1,80 @@
+import {Directive, Injectable, Optional, SkipSelf} from '@angular/core';
+
+
+/** Singleton that allows all instances of CdkAddFocusClasses to share document event listeners. */
+@Injectable()
+export class CdkFocusCauseDetector {
+  /** Whether a keydown event has just occurred. */
+  get keydownOccurred() { return this._keydownOccurred; }
+  private _keydownOccurred = false;
+
+  get mousedownOccurred() { return this._mousedownOccurred; }
+  private _mousedownOccurred = false;
+
+  constructor() {
+    document.addEventListener('keydown', () => {
+      this._keydownOccurred = true;
+      setTimeout(() => this._keydownOccurred = false, 0);
+    }, true);
+
+    document.addEventListener('mousedown', () => {
```

Comment:
A lot of touch devices actually emit a `mousedown` event for touch clicks, but I don't think that this really reliable.
> Test for `mousedown` and `touchstart` events: http://codepen.io/DevVersion/pen/qRNgOZ

In the Material 1 interaction service we had to buffer the touch events: [see here](https://github.com/angular/material/blob/master/src/core/services/interaction/interaction.js#L113)